bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitMetadataCharacteristicFactory.java

```diff
@@ -0,0 +1,295 @@
+            put(CURRENT_HEATING_COOLING_STATE,
+                    HomekitMetadataCharacteristicFactory::createCurrentHeatingCoolingStateCharacteristic);
+            put(CURRENT_VISIBILITY, HomekitMetadataCharacteristicFactory::createCurrentVisibilityCharacteristic);
+            put(IDENTIFIER, HomekitMetadataCharacteristicFactory::createIdentifierCharacteristic);
+            put(INPUT_DEVICE_TYPE, HomekitMetadataCharacteristicFactory::createInputDeviceTypeCharacteristic);
+            put(INPUT_SOURCE_TYPE, HomekitMetadataCharacteristicFactory::createInputSourceTypeCharacteristic);
+            put(NAME, HomekitMetadataCharacteristicFactory::createNameCharacteristic);
+            put(PICTURE_MODE, HomekitMetadataCharacteristicFactory::createPictureModeCharacteristic);
+            put(SLEEP_DISCOVERY_MODE, HomekitMetadataCharacteristicFactory::createSleepDiscoveryModeCharacteristic);
+            put(TARGET_HEATER_COOLER_STATE,
+                    HomekitMetadataCharacteristicFactory::createTargetHeaterCoolerStateCharacteristic);
+            put(TARGET_HEATING_COOLING_STATE,
+                    HomekitMetadataCharacteristicFactory::createTargetHeatingCoolingStateCharacteristic);
+            put(VOLUME_CONTROL_TYPE, HomekitMetadataCharacteristicFactory::createVolumeControlTypeCharacteristic);
+        }
+    };
+
+    public static Optional<Characteristic> createCharacteristic(String characteristic, Object value) {
+        var type = HomekitCharacteristicType.valueOfTag(characteristic);
+        if (type.isEmpty() || !optional.containsKey(type.get()))
+            return Optional.empty();
+        return Optional.of(optional.get(type.get()).apply(value));
+    }
+
+    private static Supplier<CompletableFuture<Integer>> getInteger(Object value) {
+        int intValue;
+        if (value instanceof BigDecimal) {
+            intValue = ((BigDecimal) value).intValue();
+        } else if (value instanceof Float) {
+            intValue = ((Float) value).intValue();
+        } else if (value instanceof Integer) {
+            intValue = (Integer) value;
+        } else if (value instanceof Long) {
+            intValue = ((Long) value).intValue();
+        } else {
+            intValue = Integer.valueOf(value.toString());
+        }
+        return () -> CompletableFuture.completedFuture(intValue);
+    }
+
+    private static Supplier<CompletableFuture<String>> getString(Object value) {
+        return () -> CompletableFuture.completedFuture(value.toString());
+    }
+
+    private static <T extends Enum<T>> Supplier<CompletableFuture<T>> getEnum(Object value, Class<T> klazz) {
+        T enumValue = Enum.valueOf(klazz, value.toString());
+        return () -> CompletableFuture.completedFuture(enumValue);
+    }
+
+    private static <T extends Enum<T>> Supplier<CompletableFuture<T>> getEnum(Object value, Class<T> klazz, T trueValue,
+            T falseValue) {
+        if (value.equals(true) || value.equals("true")) {
```

Comment:
Because depending on where the metadata came from (.items file, UI, JSR223 script setting it) the Object might be a String, or a Boolean. I'm not sure I follow what you mean by `trueValue would be "false"` - `trueValue` will be a value of a CharacteristicEnum (something like `ActiveEnum.ACTIVE`). There's no need for inversion when setting things like this from metadata, so `value` would never be `false` when you're trying to get `ActiveEnum.ACTIVE`. Also, `false.equals(true) || false.equals("true")` will return false, as will `"false".equals(true) || "false".equals("true")`.

Reply:
right. forgot that this is static and does not require mapping. 
thanks for the explanation.